File: sdk/python/log_test.py
"""Structured logging tests for the Python SDK.

No running NATS is needed - the logger writes through an injected sink.

Run: uv run --with nats-py -m unittest log_test   (from sdk/python)
"""

import json
import unittest

from aether import Logger, new_logger


def capture():
    lines: list[str] = []
    return lines, lines.append


class LogTest(unittest.TestCase):
    def test_json_carries_base_and_call_fields(self):
        lines, write = capture()
        log = Logger({"component": "thrall", "app": "counter", "name": "worker"}, level=20, fmt="json", write=write)
        log.info("thrall ready", pid=42)

        self.assertEqual(len(lines), 1)
        rec = json.loads(lines[0])
        self.assertEqual(rec["level"], "INFO")
        self.assertEqual(rec["msg"], "thrall ready")
        self.assertEqual(rec["component"], "thrall")
        self.assertEqual(rec["app"], "counter")
        self.assertEqual(rec["name"], "worker")
        self.assertEqual(rec["pid"], 42)
        self.assertIsInstance(rec["time"], str)

    def test_level_below_threshold_dropped(self):
        lines, write = capture()
        log = Logger(level=30, fmt="json", write=write)  # warn
        log.info("dropped")
        log.warn("kept")

        self.assertEqual(len(lines), 1)
        self.assertEqual(json.loads(lines[0])["msg"], "kept")

    def test_text_is_default_rendering(self):
        lines, write = capture()
        log = Logger({"name": "w"}, level=20, fmt="text", write=write)
        log.info("hello")

        self.assertFalse(lines[0].startswith("{"))
        self.assertIn("INFO", lines[0])
        self.assertIn("hello", lines[0])
        self.assertIn("name=w", lines[0])

    def test_with_derives_child_logger(self):
        lines, write = capture()
        base = Logger({"app": "counter"}, level=20, fmt="json", write=write)
        base.with_(name="worker").error("boom", err="nope")

        rec = json.loads(lines[0])
        self.assertEqual(rec["app"], "counter")
        self.assertEqual(rec["name"], "worker")
        self.assertEqual(rec["err"], "nope")
        self.assertEqual(rec["level"], "ERROR")

    def test_new_logger_factory_defaults_from_env(self):
        # new_logger() reads env; with nothing set it defaults to info/text and does not raise.
        log = new_logger(component="thrall")
        self.assertIsInstance(log, Logger)


if __name__ == "__main__":
    unittest.main()

File: sdk/ts/src/log.test.ts
import { test, expect } from "bun:test";
import { newLogger } from "./log";

// Capture output through the injectable writer so tests never touch stderr.
function capture() {
  const lines: string[] = [];
  return { lines, write: (line: string) => lines.push(line) };
}

test("json format carries base and per-call fields", () => {
  const { lines, write } = capture();
  const log = newLogger({ component: "thrall", app: "counter", name: "worker" }, { level: "info", format: "json", write });
  log.info("thrall ready", { pid: 42 });

  expect(lines.length).toBe(1);
  const rec = JSON.parse(lines[0]);
  expect(rec.level).toBe("INFO");
  expect(rec.msg).toBe("thrall ready");
  expect(rec.component).toBe("thrall");
  expect(rec.app).toBe("counter");
  expect(rec.name).toBe("worker");
  expect(rec.pid).toBe(42);
  expect(typeof rec.time).toBe("string");
});

test("level below threshold is dropped", () => {
  const { lines, write } = capture();
  const log = newLogger({}, { level: "warn", format: "json", write });
  log.info("dropped");
  log.warn("kept");

  expect(lines.length).toBe(1);
  expect(JSON.parse(lines[0]).msg).toBe("kept");
});

test("text format is the default rendering", () => {
  const { lines, write } = capture();
  const log = newLogger({ name: "w" }, { level: "info", format: "text", write });
  log.info("hello");

  expect(lines[0]).not.toStartWith("{");
  expect(lines[0]).toContain("INFO");
  expect(lines[0]).toContain("hello");
  expect(lines[0]).toContain("name=w");
});

test("with() derives a child logger merging fields", () => {
  const { lines, write } = capture();
  const base = newLogger({ app: "counter" }, { level: "info", format: "json", write });
  base.with({ name: "worker" }).error("boom", { err: "nope" });

  const rec = JSON.parse(lines[0]);
  expect(rec.app).toBe("counter");
  expect(rec.name).toBe("worker");
  expect(rec.err).toBe("nope");
  expect(rec.level).toBe("ERROR");
});
